language/move-symbol-pool/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! A global, uniqued cache of strings that is never purged.
+//!
+//! This module provides storage for strings that are meant to remain in use for
+//! the entire running duration of a program. Strings that are stored in this
+//! global, static cache are never evicted, and so the memory consumed by them
+//! can only ever grow.
+//!
+//! The strings can be accessed via the [`Symbol`] type, which acts like a
+//! pointer to the underlying cached string value.
+//!
+//! NOTE: If you're looking for a `#[forbid(unsafe_code)]` attribute here, you
```

Comment:
I'm wondering why are we re-implementing our own (unsafe) intern solution. In the Move Prover we use the [internement](https://crates.io/crates/internment) crate successfully, so what's wrong with this one?

Reply:
The main crate we considered was servo's string_cache (https://docs.rs/string_cache/0.8.1/string_cache/). The problem with it is that it is ref counted. And all of our strings we parse are likely to live for the duration of the program, so the ref counting won't buy us much.

The internment crate is a nice general wrapper around `Box::leak`, but it is not really optimized for strings. 
We want to use this eventually for strings coming out of the parser (so any name/identifier a user gives). And about ~50% of those strings in the DF+Stdlib are less than 7 characters. There is a great optimization we can use from string_cache to stuff those characters into the actual symbol instead of allocating a string in the table (the bit stuffing code from string_cache: https://github.com/servo/string-cache/blob/master/src/atom.rs#L191)

In short, the existing crates don't quite match up with what we want to do long term, and the code here is rather simple to maintain and easy to optimize as we go forward